kvserver: replace ad-hoc stats in RaftTransport by metrics

pav-kv opened the pull request:
These stats were only printed in text to the logs, and hard to use and match
with other existing metrics in monitoring.

There is a queue size metric introduced in 5d51ab34 which properly exports the
`queue` counter to monitoring. The exported graphs allow observing both the
current, and historical queue sizes (including the approx. max size), so this
PR safely removes both stats as obsolete.

This PR also similarly replaces all the remaining ad-hoc counters with proper
metrics, and removes the plumbing and data structures that were previously
maintaining those stats.

The granularity of the new metrics is per-node/transport. This is due to the
metrics infrastructure not being particularly good at handling high cardinality
combinations of labels such as O(N^2) node ID pairs.

Part of #83917

Release note: None

pav-kv committed:
kvserver: remove ad-hoc queue stats from RaftTransport

These stats are only printed in text to the logs. There is a queue size metric
introduced in 5d51ab34 which properly exports these stats to monitoring. The
exported graphs allow observing both the current, and historical queue sizes
(including the max size).

This commit removes the `queue` and `queueMax` stats as obsolete.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85951)
<!-- Reviewable:end -->

pav-kv committed:
kvserver: replace ad-hoc message counters with metrics

The counters were only printed to text logs. This commit replaces them with
proper metrics that can be used for monitoring and alerting.

Release note: None

pav-kv committed:
kvserver: replace ad-hoc response counters with metrics

The counters were only printed to text logs. This commit replaces them with
proper metrics that can be used for monitoring and alerting.

Release note: None

pav-kv committed:
kvserver: replace clientDropped stats with metrics

This counter is only printed to text logs. This commit replaces it with a
proper metric that can be used for monitoring and alerting. It also removes the
surrounding plumbing and data structures that maintained the stats.

Release note: None

pavelkalinnikov commented:
@ajwerner This removes your TODO because it's no longer relevant after this PR: we reduce the granularity of counters. Is it fine, or you would still leave/reword the TODO?

tbg reviewed (approved):
Nice!

pavelkalinnikov commented on pkg/kv/kvserver/raft_transport_metrics.go:
```diff
@@ -45,5 +47,19 @@ messages to at least one peer.`,
+
+		ResponsesSent: metric.NewCounter(metric.Metadata{
+			Name:        "raft.transport.resp-sent",
+			Help:        "Number of Raft responses sent by the Raft Transport",
```
Done.

pavelkalinnikov commented:
bors r=tbg
